kiwi/lib/search.py

```diff
@@ -0,0 +1,438 @@
+        kiwi_config['quiet'] = False
+    if kiwi_config['verbose'] is None:
+        kiwi_config['verbose'] = False
+    trainconfig = train.Configuration(**kiwi_config)
+
+    # TODO: integrate PTL trainer callback
+    # ptl_callback = PyTorchLightningPruningCallback(trial, monitor='val_PEARSON')
+    # train_info = train.run(trainconfig, ptl_callback=ptl_callback)
+
+    logger.info(f"############# STARTING TRIAL {trial.number} #############")
+    logger.info(f"PARAMETERS: {dict(zip(names, values))}")
+    for name, value in zip(names, values):
+        logger.info(f'{name}: {value}')
+
+    try:
+        train_info = train.run(trainconfig)
+    except RuntimeError as e:
+        logger.info(f'ERROR OCCURED; SKIPPING TRIAL: {e}')
+        return -1
+
+    logger.info(f"############# TRIAL {trial.number} FINISHED #############")
+    result = train_info.best_metrics.get(main_metric, -1)
+    if result == -1:
+        logger.error(
+            'Trial did not converge:\n'
+            f'    `train_info.best_metrics={train_info.best_metrics}`\n'
+            f'Setting result to -1'
+        )
+
+    logger.info(f"RESULTS: {result} {main_metric}")
+    logger.info(f"PARAMETERS: {dict(zip(names, values))}")
+    for name, value in zip(names, values):
+        logger.info(f'{name}: {value}')
+
+    return result
+
+
+def setup_run(directory: Path, seed: int, debug=False, quiet=False) -> Path:
+    # TODO: replace all of this with the MLFlow integration: optuna.integration.mlflow
+    #   In particular: let the output directory be created by MLflow entirely
+    if not directory.exists():
+        # Initialize a new folder with name '0'
+        output_dir = directory / '0'
+        output_dir.mkdir(parents=True)
+    else:
+        # Initialize a new folder incrementing folder count
+        output_dir = directory / str(len(list(directory.glob('*'))))
+        if output_dir.exists():
+            backup_directory = output_dir.with_name(
+                f'{output_dir.name}_{datetime.now().isoformat()}'
+            )
+            logger.warning(
+                f'Folder {output_dir} already exists; moving it to {backup_directory}'
+            )
+        output_dir.mkdir(parents=True)
+
+    logger.info(f'Initializing new search folder at: {output_dir}')
+
+    configure_logging(output_dir=output_dir, verbose=debug, quiet=quiet)
+    configure_seed(seed)
+
+    return output_dir
+
+
+def run(config: Configuration):
+    try:
+        import optuna
+    except ImportError as e:
+        logger.error(
+            f'ImportError: {e}. Install the search dependencies with '
+            '``pip install -U openkiwi[search]``, or with ``poetry install -E search`` '
+            'when setting up for local development'
+        )
+        exit()
+
+    output_dir = setup_run(config.directory, config.seed)
+
+    logger.info(f'Saving all Optuna results to: {output_dir}')
+
+    # FIXME: this is not neat (but it does work)
```

Comment:
Add a comment explaining why you do this so it becomes less of black-magic. (you can also leave the fix-me)